File: PHP/Patterns/src/Behavioral/Observer/OnlineMarketPlace.php
<?php

namespace DesignPatterns\Behavioral\Observer;

use DesignPatterns\Behavioral\Observer\Enums\EventType;
use DesignPatterns\Behavioral\Observer\Interfaces\Subscriber;
use DesignPatterns\Behavioral\Observer\Models\Offer;
use DesignPatterns\Behavioral\Observer\Models\Product;

class OnlineMarketPlace
{
    private array $subscribers;
    protected array $products;
    protected array $offers;

    public function __construct()
    {
        $this->subscribers = [];
        $this->products = [];
        $this->offers = [];
        $this->initSubscribersEvents();
    }

    /**
     * @return void
     */
    private function initSubscribersEvents(): void
    {
        $this->subscribers[EventType::OPEN_POSITION->name] = [];
        $this->subscribers[EventType::NEW_PRODUCT->name] = [];
        $this->subscribers[EventType::NEW_OFFER->name] = [];
    }

/**
 * @param EventType $eventType
 * @param Subscriber $subscriber
 * @return void
 */

    public function subscribe(EventType $eventType, Subscriber $subscriber): void
    {
        $this->subscribers[$eventType->name][] = $subscriber;
    }

    /**
     * @param EventType $eventType
     * @param Subscriber $subscriber
     * @return void
     */
    public function unsubscribe(EventType $eventType, Subscriber $subscriber): void
    {
        $this->subscribers[$eventType->name] = array_filter($this->subscribers[$eventType->name], function ($sub) use ($subscriber) {
            return $sub->getName() !== $subscriber->getName();
        });

    }

    /**
     * @param Product $product
     * @return void
     */
    public function addProduct(Product $product): void
    {
        $this->products[] = $product;
        $this->notifySubscribers(EventType::NEW_PRODUCT, "New product is added " . $product->getName());

    }

    /**
     * @param Offer $offer
     * @return void
     */
    public function addOffer(Offer $offer): void
    {
        $this->offers[] = $offer;
        $this->notifySubscribers(EventType::NEW_OFFER, "New offer added " . $offer->getMessage());

    }

    public function addNewJobOpening(string $jobOpeningMessage): void
    {
        $this->notifySubscribers(EventType::OPEN_POSITION, "New Opening Position is available" . $jobOpeningMessage);
    }

    /**
     * @param EventType $eventType
     * @param string $message
     * @return void
     */
    private function notifySubscribers(EventType $eventType, string $message): void
    {
        foreach ($this->subscribers[$eventType->name] as $subscriber) {
            $subscriber->notify($message);
        }
    }

}

File: dart_lang/lib/b_behavioral/a_observer/b_enhanced_solution/online_marketplace.dart
import 'package:dart_lang/b_behavioral/a_observer/b_enhanced_solution/observer.dart';

class OnlineMarketPlace {
  Map<EventType, List<Subscriber>> _subscribers;
  List<Product> _products;
  List<Offer> _offers;

  OnlineMarketPlace()
      : _subscribers = {},
        _products = [],
        _offers = [] {
    initSubscribers();
  }

  void initSubscribers() {
    for (EventType type in EventType.values) {
      _subscribers[type] = [];
    }
  }

  void subscribe(Subscriber subscriber, EventType type) {
    _subscribers[type]!.add(subscriber);
  }

  void unSubscribe(Subscriber subscriber, EventType type) {
    _subscribers[type]!.remove(subscriber);
  }

  void addProduct(Product product) {
    _products.add(product);
    notifySubscribers(
        EventType.newProduct, "New product is added ${product.name}");
  }

  void addOffer(Offer offer) {
    _offers.add(offer);
    notifySubscribers(
        EventType.newOffer, "New offer is added: ${offer.message}");
  }

  void notifySubscribers(EventType type, String message) {
    for (Subscriber subscriber in _subscribers[type]!) {
      subscriber.notify(message);
    }
  }
}
